R/fair-aaa.R

```diff
@@ -0,0 +1,137 @@
+#' [equal_opportunity()], or [equalized_odds()] to learn more about how the
+#' output of this function can be used.
+#'
+#' @examples
+#' data(hpc_cv)
+#'
+#' # `demographic_parity`, among other fairness metrics,
+#' # is generated with `fairness_metric()`:
+#' diff_range <- function(x, ...) {diff(range(x$.estimate))}
+#' demographic_parity_ <-
+#'   fairness_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "demographic_parity",
+#'     .post = diff_range
+#'   )
+#'
+#' m_set <- metric_set(demographic_parity_(Resample))
+#'
+#' m_set(hpc_cv, truth = obs, estimate = pred)
+#'
+#' # the `post` argument can be used to accommodate a wide
+#' # variety of parameterizations. to encode demographic
+#' # parity as a ratio inside of a difference, for example:
+#' ratio_range <- function(x, ...) {
+#'   range <- range(x$.estimate)
+#'   range[1] / range[2]
+#' }
+#'
+#' demographic_parity_ratio <-
+#'   fairness_metric(
+#'     .fn = detection_prevalence,
+#'     .name = "demographic_parity_ratio",
+#'     .post = ratio_range
+#'   )
+#'
+#' @export
+fairness_metric <- function(.fn, .name, .post) {
+  if (rlang::is_missing(.fn) || !inherits_any(.fn, c("metric", "metric_set"))) {
+    rlang::abort("`.fn` must be a metric function or metric set.")
+  }
+  if (rlang::is_missing(.name) || !is_string(.name)) {
+    abort("`.name` must be a string.")
+  }
+  if (rlang::is_missing(.post) || !is_function(.post)) {
+    abort("`.post` must be a function.")
+  }
+
+  function(by) {
+    by_str <- rlang::as_string(rlang::enexpr(by))
+    res <-
+      function(data, ...) {
+        gp_vars <- dplyr::group_vars(data)
+
+        res <- dplyr::group_by(data, {{by}}, .add = TRUE)
+        res <- .fn(res, ...)
+
+        if (length(gp_vars) > 0) {
+          splits <- vctrs::vec_split(res, res[gp_vars])
+          .estimate <- vapply(splits$val, .post, numeric(1), ...)
+        } else {
+          .estimate <- .post(res, ...)
+        }
+
+        if (!rlang::is_bare_numeric(.estimate)) {
+          abort(
+            "`.post` must return a single numeric value.",
+            call = rlang::call2("fairness_metric")
+          )
+        }
+
+        if (length(gp_vars) > 0) {
+          res <- dplyr::group_by(res, !!!dplyr::groups(data), .add = FALSE)
+        }
+
+        res <-
+          dplyr::summarize(
+            res,
+            .metric = .name,
+            !!".by" := by_str,
+            .estimator = .estimator[1],
+            .groups = "drop"
+          )
+
+        res$.estimate <- .estimate
+
+        res
+      }
+    res <- new_class_metric(res, direction = "minimize")
+    attr(res, "by") <- by_str
+    res
+  }
+}
+
+diff_range <- function(x, ...) {
+  diff(range(x$.estimate))
+}
+
+max_positive_rate_diff <- function(x, ...) {
+  metric_values <- vctrs::vec_split(x, x$.metric)
+
+  positive_rate_diff <- vapply(metric_values$val, diff_range, numeric(1), ...)
+
+  max(positive_rate_diff)
+}
```

Comment:
Unless you are expecting these function to be used across many new metrics, I would move them closer to where there are used.

Reply:
`diff_range()` is just used in two places, though `max_positive_rate_diff()` just one. I wasn't sure whether we would eventually want to supply a larger suite of fairness metrics than these three—if we feel good about just supplying these canonical ones, `max_positive_rate_diff()` should definitely move to `equalized_odds()`!

Reply:
Addressed in [82717fa](https://github.com/tidymodels/yardstick/pull/434/commits/82717fadf5409b9630ed02c4ba95aa1e8e40fcee). :)